tzrec/modules/backbone_module.py

```diff
@@ -0,0 +1,141 @@
+
+    Output shape:
+        - 2D tensor with shape: ``(batch_size, 1)``
+    """
+
+    def __init__(
+        self, use_variant: bool = False, l2_regularization: float = 1e-4
+    ) -> None:
+        super().__init__()
+        self.use_variant = use_variant
+        self.l2_regularization = l2_regularization
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """Forward pass of FM module.
+
+        Args:
+            inputs: 3D tensor with shape (batch_size, field_size, embedding_size) # NOQA
+
+        Returns:
+            torch.Tensor: FM interaction output with shape (batch_size, 1)
+        """
+        # Note: Dimension validation is skipped for FX tracing compatibility
+        # Users should ensure inputs are 3D tensors with shape (batch_size, field_size, embedding_size) # NOQA
+
+        feature = inputs
+        batch_size, field_size, embedding_size = feature.shape
+
+        if self.use_variant:
+            # Variant FM: more computationally efficient for sparse features
+            # Sum pooling across fields
+            sum_of_features = torch.sum(feature, dim=1)  # (batch_size, embedding_size)
+            square_of_sum = sum_of_features.pow(2)  # (batch_size, embedding_size)
+
+            # Sum of squares
+            sum_of_squares = torch.sum(
+                feature.pow(2), dim=1
+            )  # (batch_size, embedding_size)
+
+            # FM interaction: 0.5 * (square_of_sum - sum_of_squares)
+            fm_output = 0.5 * (
+                square_of_sum - sum_of_squares
+            )  # (batch_size, embedding_size)
+
+            # Sum across embedding dimension and add batch dimension
+            output = torch.sum(fm_output, dim=1, keepdim=True)  # (batch_size, 1)
+        else:
+            # Standard FM computation using vectorized operations
+            # This is equivalent to pairwise interactions but FX-trace friendly
+
+            # Sum pooling across fields
+            sum_of_features = torch.sum(feature, dim=1)  # (batch_size, embedding_size)
+            square_of_sum = sum_of_features.pow(2)  # (batch_size, embedding_size)
+
+            # Sum of squares
+            sum_of_squares = torch.sum(
+                feature.pow(2), dim=1
+            )  # (batch_size, embedding_size)
+
+            # FM interaction: 0.5 * (square_of_sum - sum_of_squares)
+            fm_interaction = 0.5 * (
+                square_of_sum - sum_of_squares
+            )  # (batch_size, embedding_size)
+
+            # Sum across embedding dimension to get final output
+            output = torch.sum(fm_interaction, dim=1, keepdim=True)  # (batch_size, 1)
+
+        # Apply L2 regularization if specified (add to loss during training)
+        if self.training and self.l2_regularization > 0:
+            # Store L2 regularization term for potential use in loss calculation
+            self.l2_reg_loss = self.l2_regularization * torch.sum(feature.pow(2))
```

Comment:
We should not do L2 reg in forward. Instead, it's more effective to apply weight decay directly within the optimizer settings.